File: Segunda Practica Auxiliatura /Herencia/Eje3/Eje3.java
import java.time.LocalDate;
import java.time.Period;
import java.util.ArrayList;

class Persona {
    protected String ci;
    protected String nombre;
    protected String apellido;
    protected String celular;
    protected LocalDate fechaNac;

    public Persona() {
        this("0000", "SinNombre", "SinApellido", "00000000", LocalDate.of(2000, 1, 1));
    }

    public Persona(String ci, String nombre, String apellido, String celular, LocalDate fechaNac) {
        this.ci = ci;
        this.nombre = nombre;
        this.apellido = apellido;
        this.celular = celular;
        this.fechaNac = fechaNac;
    }

    public int calcularEdad() {
        return Period.between(fechaNac, LocalDate.now()).getYears();
    }

    public void mostrar() {
        System.out.println("CI: " + ci);
        System.out.println("Nombre: " + nombre);
        System.out.println("Apellido: " + apellido);
        System.out.println("Celular: " + celular);
        System.out.println("Fecha de Nacimiento: " + fechaNac);
    }
}

class Estudiante extends Persona {
    private String ru;
    private LocalDate fechaIngreso;
    private int semestre;

    public Estudiante() {
        this("0000", "SinNombre", "SinApellido", "00000000", LocalDate.of(2000,1,1),
             "RU0000", LocalDate.of(2023, 1, 1), 1);
    }

    public Estudiante(String ci, String nombre, String apellido, String celular, LocalDate fechaNac,
                      String ru, LocalDate fechaIngreso, int semestre) {
        super(ci, nombre, apellido, celular, fechaNac);
        this.ru = ru;
        this.fechaIngreso = fechaIngreso;
        this.semestre = semestre;
    }

    @Override
    public void mostrar() {
        super.mostrar();
        System.out.println("RU: " + ru);
        System.out.println("Fecha de Ingreso: " + fechaIngreso);
        System.out.println("Semestre: " + semestre);
    }
}

class Docente extends Persona {
    private String nit;
    private String profesion;
    private String especialidad;
    private String sexo; 

    public Docente() {
        this("0000", "SinNombre", "SinApellido", "00000000", LocalDate.of(2000,1,1),
             "NIT0000", "Ninguna", "General", "M");
    }

    public Docente(String ci, String nombre, String apellido, String celular, LocalDate fechaNac,
                   String nit, String profesion, String especialidad, String sexo) {
        super(ci, nombre, apellido, celular, fechaNac);
        this.nit = nit;
        this.profesion = profesion;
        this.especialidad = especialidad;
        this.sexo = sexo;
    }

    public String getProfesion() {
        return profesion;
    }

    public String getSexo() {
        return sexo;
    }

    @Override
    public void mostrar() {
        super.mostrar();
        System.out.println("NIT: " + nit);
        System.out.println("Profesión: " + profesion);
        System.out.println("Especialidad: " + especialidad);
        System.out.println("Sexo: " + sexo);
    }
}

public class Main {
    public static void main(String[] args) {
        // INCISO B
        ArrayList<Estudiante> estudiantes = new ArrayList<>();
        estudiantes.add(new Estudiante("123", "Luis", "Perez", "78945612", LocalDate.of(1998, 5, 21),
                                      "RU001", LocalDate.of(2021,1,1), 6));
        estudiantes.add(new Estudiante("124", "Maria", "Lopez", "78945613", LocalDate.of(2002, 8, 15),
                                      "RU002", LocalDate.of(2022,2,1), 4));

        ArrayList<Docente> docentes = new ArrayList<>();
        docentes.add(new Docente("200", "Juan", "Garcia", "75683912", LocalDate.of(1980, 4, 11),
                                 "NIT01", "Ingeniero", "Software", "M"));
        docentes.add(new Docente("201", "Lucia", "Lopez", "75683913", LocalDate.of(1985, 6, 23),
                                 "NIT02", "Arquitecta", "Diseño", "F"));

        // INCISO C
        System.out.println("Estudiantes mayores de 25 años:");
        for (Estudiante e : estudiantes) {
            if (e.calcularEdad() > 25) {
                e.mostrar();
                System.out.println("Edad: " + e.calcularEdad());
                System.out.println("------------");
            }
        }

        // INCISO D
        Docente mayorIngeniero = null;
        for (Docente d : docentes) {
            if (d.getProfesion().equalsIgnoreCase("Ingeniero") && d.getSexo().equalsIgnoreCase("M")) {
                if (mayorIngeniero == null || d.calcularEdad() > mayorIngeniero.calcularEdad()) {
                    mayorIngeniero = d;
                }
            }
        }

        System.out.println("Docente varón con profesión 'Ingeniero' y mayor edad:");
        if (mayorIngeniero != null) {
            mayorIngeniero.mostrar();
            System.out.println("Edad: " + mayorIngeniero.calcularEdad());
        }
        System.out.println("------------");

        // INCISO E
        System.out.println("Estudiantes y docentes con el mismo apellido:");
        for (Estudiante e : estudiantes) {
            for (Docente d : docentes) {
                if (e.apellido.equalsIgnoreCase(d.apellido)) {
                    System.out.println("Apellido coincidente: " + e.apellido);
                    System.out.println("Estudiante:");
                    e.mostrar();
                    System.out.println("Docente:");
                    d.mostrar();
                    System.out.println("------------");
                }
            }
        }
    }
}

File: Segunda Practica Auxiliatura /Herencia/Eje3/Eje3.py
from datetime import datetime

# INCISO A

class Persona:
    def __init__(self, ci="0000", nombre="SinNombre", apellido="SinApellido", celular="00000000", fecha_Nac="2000-01-01"):
        self.ci = ci
        self.nombre = nombre
        self.apellido = apellido
        self.celular = celular
        self.fecha_Nac = fecha_Nac  

    def calcular_edad(self):
        anio_nac = int(self.fecha_Nac.split("-")[0])
        anio_actual = datetime.now().year
        return anio_actual - anio_nac

    def mostrar(self):
        print(f"CI: {self.ci}")
        print(f"Nombre: {self.nombre}")
        print(f"Apellido: {self.apellido}")
        print(f"Celular: {self.celular}")
        print(f"Fecha de Nacimiento: {self.fecha_Nac}")

class Estudiante(Persona):
    def __init__(self, ci="0000", nombre="SinNombre", apellido="SinApellido", celular="00000000", fecha_Nac="2000-01-01",
                 ru="RU0000", fecha_Ingreso="2023-01-01", semestre=1):
        super().__init__(ci, nombre, apellido, celular, fecha_Nac)
        self.ru = ru
        self.fecha_Ingreso = fecha_Ingreso
        self.semestre = semestre

    def mostrar(self):
        super().mostrar()
        print(f"RU: {self.ru}")
        print(f"Fecha de Ingreso: {self.fecha_Ingreso}")
        print(f"Semestre: {self.semestre}")

class Docente(Persona):
    def __init__(self, ci="0000", nombre="SinNombre", apellido="SinApellido", celular="00000000", fecha_Nac="2000-01-01",
                 nit="NIT0000", profesion="Ninguna", especialidad="General", sexo="M"):
        super().__init__(ci, nombre, apellido, celular, fecha_Nac)
        self.nit = nit
        self.profesion = profesion
        self.especialidad = especialidad
        self.sexo = sexo  

    def mostrar(self):
        super().mostrar()
        print(f"NIT: {self.nit}")
        print(f"Profesión: {self.profesion}")
        print(f"Especialidad: {self.especialidad}")
        print(f"Sexo: {self.sexo}")


# INCISO B

estudiantes = [
    Estudiante("123", "Luis", "Perez", "78945612", "1998-05-21", "RU001", "2021-01-01", 6),
    Estudiante("124", "Maria", "Lopez", "78945613", "2002-08-15", "RU002", "2022-02-01", 4),
]

docentes = [
    Docente("200", "Juan", "Garcia", "75683912", "1980-04-11", "NIT01", "Ingeniero", "Software", "M"),
    Docente("201", "Lucia", "Lopez", "75683913", "1985-06-23", "NIT02", "Arquitecta", "Diseño", "F"),
]


# INCISO C

print("Estudiantes mayores de 25 años:")
for e in estudiantes:
    if e.calcular_edad() > 25:
        e.mostrar()
        print(f"Edad: {e.calcular_edad()}")
        print("------------")


# INCISO D

mayor_ingeniero = None
for d in docentes:
    if d.profesion.lower() == "ingeniero" and d.sexo.upper() == "M":
        if mayor_ingeniero is None or d.calcular_edad() > mayor_ingeniero.calcular_edad():
            mayor_ingeniero = d

print("Docente varón con profesión 'Ingeniero' y mayor edad:")
if mayor_ingeniero:
    mayor_ingeniero.mostrar()
    print(f"Edad: {mayor_ingeniero.calcular_edad()}")
print("------------")


# INCISO E

print("Estudiantes y docentes con el mismo apellido:")
for e in estudiantes:
    for d in docentes:
        if e.apellido.lower() == d.apellido.lower():
            print(f"Apellido coincidente: {e.apellido}")
            print("Estudiante:")
            e.mostrar()
            print("Docente:")
            d.mostrar()
            print("------------")
